src/authentication.ts

```diff
@@ -0,0 +1,189 @@
+ * Stores the refresh token, access token and it's update time in session storage.
+ * Returns a promise that resolves with the access token.
+ */
+async function refreshAccessToken(refreshToken: string, redirectUrl: string) {
+  const tokenUrl = oauthUrl + "/token";
+  const payload = new URLSearchParams();
+  payload.set("grant_type", "refresh_token");
+  payload.set("refresh_token", refreshToken);
+  payload.set("redirect_uri", redirectUrl);
+  payload.set("client_id", clientId);
+  payload.set("client_secret", clientSecret);
+
+  const json = await fetchJson<RefreshTokenResponse>(tokenUrl, {
+    method: "POST",
+    body: payload
+  });
+
+  window.localStorage.setItem("access_token", json.access_token);
+  window.localStorage.setItem(
+    "access_token_expires",
+    addSeconds(Date.now(), json.expires_in).toString()
+  );
+
+  return json.access_token;
+}
+
+/**
+ * Fetches and returns parsed json.
+ * Throws error if response is not ok.
+ */
+async function fetchJson<TData>(
+  input: RequestInfo,
+  init?: RequestInit | undefined
+) {
+  const response = await fetch(input, init);
+  if (response.ok) {
+    const json: TData = await response.json();
+    return json;
+  } else {
+    throw new Error(response.statusText);
+  }
+}
+
+interface AuthorizationCodeResponse {
+  access_token: string;
+  expires_in: number;
+  refresh_token: string;
+  token_type: string;
+}
+
+interface RefreshTokenResponse {
+  access_token: string;
+  expires_in: number;
+  token_type: string;
+}
+
+/**
+ * Redirects the user to the Portal authorize page,
+ * or to the login page if the user is not logged in.
+ */
+function login() {
+  const redirectUrl = document.location.href.split("?")[0];
+
+  const params = new URLSearchParams();
+  params.set("response_type", "code");
+  params.set("client_id", clientId);
+  params.set("redirect_uri", redirectUrl);
+  params.set("scope", scope);
+  params.set("client_secret", clientSecret);
+
+  document.location.href = `${oauthUrl}/authorize?${params}`;
+}
+
+/**
+ * Logs out the user by clearing all tokens from local storage and reloading the page.
+ */
+function logout() {
+  window.localStorage.removeItem("refresh_token");
+  window.localStorage.removeItem("access_token");
+  window.localStorage.removeItem("access_token_expires");
+  window.location.reload();
+}
+
+/**
+ * Shows login button which takes the user to the authorize page.
+ */
+export function showLoginButton() {
+  const loginButton = document.createElement("button");
+  loginButton.innerText = "Log in";
+  loginButton.setAttribute("class", "login-button");
+  loginButton.onclick = ev => login();
+
+  document.body.appendChild(loginButton);
+}
+
+/**
+ * Adds an element to the body that shows the currently logged in user and a logout link.
+ */
+export function showUserInfo(accessToken: string) {
+  const jwt: User = jwtDecode(accessToken);
+
+  const userInfo = document.createElement("div");
+  userInfo.innerText = jwt.unique_name;
```

Comment:
Maybe use `userInfo.innerHTML = "<pre>" + JSON.stringify(jwt, null, 2); + "</pre>" ` to show all JWT claims

Reply:
Maybe in separate pull request.